src/promql/src/extension_plan/scalar_calculate.rs

```diff
@@ -0,0 +1,447 @@
+        let tag_indices = self
+            .tag_columns
+            .iter()
+            .map(|tag| {
+                schema
+                    .column_with_name(tag)
+                    .unwrap_or_else(|| panic!("tag column not found {tag}"))
+                    .0
+            })
+            .collect();
+        Ok(Box::pin(ScalarCalculateStream {
+            schema: self.schema.clone(),
+            project_index: self.project_index,
+            metric: baseline_metric,
+            tag_indices,
+            input,
+            have_multi_series: false,
+            done: false,
+            batch: None,
+            tag_value: None,
+        }))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metric.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        Statistics {
+            num_rows: None,
+            total_byte_size: None,
+            column_statistics: None,
+            is_exact: false,
+        }
+    }
+}
+
+impl DisplayAs for ScalarCalculateExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(f, "ScalarCalculateExec: tags={:?}", self.tag_columns)
+            }
+        }
+    }
+}
+
+pub struct ScalarCalculateStream {
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    metric: BaselineMetrics,
+    tag_indices: Vec<usize>,
+    project_index: (usize, usize),
+    have_multi_series: bool,
+    done: bool,
+    batch: Option<RecordBatch>,
+    tag_value: Option<Vec<String>>,
+}
+
+impl RecordBatchStream for ScalarCalculateStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+impl ScalarCalculateStream {
+    fn update_batch(&mut self, batch: RecordBatch) -> DataFusionResult<()> {
+        let _timer = self.metric.elapsed_compute();
+        // if have multi time series, scalar will return NaN
+        if self.have_multi_series {
+            return Ok(());
+        }
+        // fast path: no tag columns means all data belongs to the same series.
+        if self.tag_indices.is_empty() {
+            self.append_batch(batch)?;
+            return Ok(());
+        }
+        let all_same = |val: Option<&str>, array: &StringArray| -> bool {
+            if let Some(v) = val {
+                array.iter().all(|s| s == Some(v))
+            } else {
+                array.is_empty() || array.iter().skip(1).all(|s| s == Some(array.value(0)))
+            }
+        };
+        // assert the entire batch belong to the same series
+        let all_tag_columns_same = if let Some(tags) = &self.tag_value {
+            tags.iter()
+                .zip(self.tag_indices.iter())
+                .all(|(value, index)| {
+                    let array = batch.column(*index);
+                    let string_array = array.as_any().downcast_ref::<StringArray>().unwrap();
```

Comment:
So, we still assume that the tag is always the string type? I think it's better to return an error instead of panic. What do you think? @waynexia

Reply:
I simply follow the assumption in current planner, in current planner, all tag column are consider strings, non-string tag columns will lead panic

Reply:
Got it. I think we should support other data types for tags cc @waynexia

Reply:
Other places have checked the type, but to be self-contained, I prefer to check it again in this plan.